src/cljam/io/vcf/util/validator.clj

```diff
@@ -0,0 +1,335 @@
+(ns cljam.io.vcf.util.validator
+  (:require [cljam.io.vcf.util :as vcf.util]))
+
+(defrecord VCFValidator [ploidy file-type defs samples])
+
+(defn- error [path msg & args]
+  {:errors {path [(apply format msg args)]}})
+
+(defn- warn [path msg & args]
+  {:warnings {path [(apply format msg args)]}})
+
+(defn- merge-validation-results
+  ([] nil)
+  ([res] res)
+  ([res1 res2]
+   (letfn [(rec [x y]
+             (cond (nil? x) y
+                   (nil? y) x
+                   (map? x) (merge-with rec x y)
+                   :else (into x y)))]
+     (rec res1 res2)))
+  ([res1 res2 res3 & more]
+   (reduce merge-validation-results res1 (list* res2 res3 more))))
+
+(defn- prepend-keys [res keys]
+  ;; takes {:errors {[:chr] ["foo"]}}
+  ;; returns {:errors {[key ... :chr] ["foo"]}}
+  (when-not (nil? res)
+    (into {}
+          (map (fn [[kind m]]
+                 [kind
+                  (into {} (map (fn [[path messages]]
+                                  [(into keys path) messages]))
+                        m)]))
+          res)))
+
+(defn- error-on-bcf [validator path & args]
+  (let [f (if (= :bcf (:file-type validator)) error warn)]
+    (apply f path args)))
+
+(defn- prep-contig-defs [_validator contigs]
+  (into {} (map (juxt :id identity)) contigs))
+
+(defn- validate-contig [validator {:keys [chr]}]
+  (if (and (string? chr) (re-matches #"[^<>, \t]+" chr))
+    (when-not (get-in validator [:defs :contig chr])
+      (error-on-bcf validator [:chr] "Contig %s not defined in meta info" chr))
+    (error [:chr]
+           (str ":chr must be a non-empty string that consists of characters "
+                "other than whitespace, commas or angle brackets, but got "
+                (pr-str chr)))))
+
+(defn- validate-pos [_validator {:keys [pos]}]
+  (when-not (integer? pos)
+    (error [:pos] (str ":pos must be integer, but got " (pr-str pos)))))
+
+(defn- validate-ref [_validator {r :ref}]
+  (when-not (and (string? r) (re-matches #"(?i)[ATGCN]+" r))
+    (error [:ref]
+           (str ":ref must be a non-empty string consisting of A/C/G/T/N "
+                "(case insensitive), but got " (pr-str r)))))
+
+(defn- validate-alt [_validator {:keys [alt] :as variant}]
+  (if (or (nil? alt) (sequential? alt))
+    (some->> alt
+             (keep-indexed
+              (fn [i allele]
+                (cond
+                  (and (not (nil? allele))
+                       (or (not (string? allele)) (= allele "")))
+                  {[:alt i]
+                   [(str "Every allele in :alt must be a non-empty string or nil, "
+                         "but got " (pr-str allele))]}
+
+                  (and (string? (:ref variant))
+                       (= :other (:type (vcf.util/inspect-allele (:ref variant) allele))))
+                  {[:alt i] [(str "Unexpected allele found: " allele)]})))
```

Comment:
If the [IUPAC ambiguity codes](https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L229-L234) are used, it will result in an error. Is there no current need to consider this?

Reply:
I'm not sure to what extent cljam supports IUPAC ambiguity codes or symbolic alleles in general, but I don’t think it would report any error in this particular case.

```clojure
(vcf.util/inspect-allele "T" "<R>")
;=> {:type :id, :id "R"}
```

Do you have any concrete example that produces an error?

Reply:
I think that according to the specification, it should be allowed to write the IUPAC code defined in `##ALT` in the `ALT` field.
For example:
```
##ALT=<ID=R,Description="IUPAC code R = A/G">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO
chr1      1     .       C       R       .       PASS    .
```
In other words, I'm concerned that there may be situations where an alphabet other than `ACGTN` should be accepted. For instance, when using the `vcf.util/inspect-allele [ref alt]` function, as demonstrated in `(vcf.util/inspect-allele "T" "R")`, it might need to handle such alphabets in the alt parameter.
However, looking at the [VCFv4.4](https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L2620), I'm wondering if the use of IUPAC ambiguity codes is not prohibited, but not recommended :
> IUPAC ambiguity codes should be converted to a concrete base.

Reply:
From my understanding, IUPAC ambiguity codes are merely an example use of symbolic alleles.

The spec describes `ALT` as follows, which indicates IDs of symbolic alleles are intended to be used being surrounded in brackets, not to be used as bare IDs:

> Each allele in this list must be one of:
> - a non-empty String of bases (A,C,G,T,N; case insensitive);
> - the `*` symbol (allele missing due to overlapping deletion);
> - the MISSING value `.` (no variant);
> - an angle-bracketed ID String (`<ID>`);
> - the unspecified allele `<*>` as described in Section 5.5; or
> - a breakend replacement string as described in Section 5.4
>
> https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L361

Also, the description below is for `REF`, not for `ALT`, and considering the other context, I think it's about how to resolve symbolic allele abuse in `REF`.

> IUPAC ambiguity codes should be converted to a concrete base.

Reply:
Your explanation helped me to understand correctly. Thank you for your explanation 👍